src/main/java/com/github/princesslana/eriscasper/gateway/Gateway.java

```diff
@@ -85,14 +88,31 @@ private void sequenceNumberSeen(Optional<SequenceNumber> seq) {
   }
 
   @SuppressWarnings("unchecked")
-  public Observable<Event> connect(String url, BotToken token) {
+  public Observable<Event> connect(String url, BotToken token, Optional<Shard> shard) {
     Observable<Payload> ps =
         ws.connect(String.format("%s?v=%s&encoding=%s", url, VERSION, ENCODING))
             .ofType(RxWebSocketEvent.StringMessage.class)
             .map(RxWebSocketEvent.StringMessage::getText)
             .flatMapMaybe(
                 Singles.toMaybeAnd(
                     payloads::read, (s, t) -> LOG.warn("Error reading payload: {}", s, t)))
+            .takeUntil(
+                payload -> {
+                  if (payload.op() == OpCode.INVALID_SESSION) {
+                    if (payload.d().map(JsonNode::asBoolean).orElse(false)) {
+                      resume(ws, token);
+                    } else {
+                      // blocking since we shouldn't be doing anything if this happens
+                      // (this is an invalid session which can't be resumed.
+                      Completable.fromAction(() -> ws.close(1002, "Invalid session."))
+                          .doOnComplete(
+                              () -> LOG.warn("Socket disconnected due to an invalid session."))
+                          .blockingAwait();
```

Comment:
We shouldn't need to use something blocking during an observable stream.

Reply:
It's blocking because we don't want anything more to even try to come through, if we do it will throw broken pipes and have issues with the websocket. If we block it then we can avoid anything even attempting to come through when it's a non-resumable invalid session.

Reply:
Putting the blocking here isn't stopping other websocket events from coming through or happening.

Reply:
Alright, how would we push through that command into the stream? Should we just run it outside of a completable.

Reply:
I'm not really sure, this isn't something I have thought about. I'm not 100% clear on what situations lead to this happening and what action we're taking.

Reply:
I'm not exactly sure what _can_ throw a non-resumable invalid sessions in all honesty. I tried for around 3 hours to get github to throw me an invalid session without success. Invalid tokens throw an initial unauthorized message not an invalid session message so that's a bit odd. Basically right now within this codebase we're just shutting down the websocket.

Reply:
Ok, so I'm not keen to merge something where we're not sure what we're meant to be doing.

Reply:
Alright, maybe we should remove it then, or just throw a fatal exception when I receive an invalid session which can't be resumed. Most invalid sessions are able to be resumed and I haven't seen one where I couldn't. So maybe that's the right move? Not 100% sure.

Reply:
I'd say remove anything related to #64 from this PR and address it at another time.

Reply:
Sounds good to me. I think it's a more complicated issue due to being able to test and understand the circumstances that lead to the outcomes we're trying to prevent.